installer/src/App.jsx

```diff
@@ -1,147 +1,395 @@
+    ]);
+    const output = await command.execute();
+    const boardList = JSON.parse(output.stdout);
+    return boardList
+        .filter((row) =>
+            row.matching_boards?.find((o) => fqbnList.includes(o.fqbn))
+        )
+        .map((board) => ({
+            name: board.matching_boards[0]?.name,
+            fqbn: board.matching_boards[0]?.fqbn,
+            port: board.port?.address,
+        }));
+};
 
 function App() {
-  const [greetMsg, setGreetMsg] = useState("");
-  const [version, setVersion] = useState("")
-  const [name, setName] = useState("");
-  const [boards, setBoards] = useState([])
-  const [selectedBoard, setSelectedBoard] = useState("")
-  const [uploading, setUploading] = useState(false);
-
-  async function greet() {
-    // Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
-    setGreetMsg(await invoke("greet", { name }));
-  }
-
-  useEffect(() => {
-    const getVersion = async () => {
-      const command = Command.sidecar('../resources/arduino-cli/arduino-cli', [
-        'version',
-        '--format',
-        'json'
-      ])
-      const output = await command.execute()
-      const versionObject = JSON.parse(output.stdout)
-      setVersion(`${versionObject.VersionString} - ${versionObject.Date}`)
-    }
+    // these properties are not part of the "internal state" of the component
+    const [version, setVersion] = useState('');
+    const [name, setName] = useState('');
+    const [leftServo, setLeftServo] = useState('3');
+    const [rightServo, setRightServo] = useState('4');
 
-    getVersion().catch(console.error)
+    const textAreaRef = useRef(null);
+    const initialState = {
+        stdout: '', // stdout as provided by the arduino-cli
+        code: null, // code as provided by the arduino-cli
+        boards: [],
+        selectedBoard: null,
+        uploading: false,
+        uploadButtonEnabled: false,
+        error: null,
+    };
 
-  }, [])
+    const [state, dispatch] = useReducer(reducer, initialState);
 
-  useEffect(() => {
-    const getBoards = async (fqbn_list) => {
-      const command = Command.sidecar('../resources/arduino-cli/arduino-cli', [
-        'board',
-        'list',
-        '--format',
-        'json'
-      ])
-      const output = await command.execute()
-      const boardList = JSON.parse(output.stdout)
+    const handleError = (error) => {
+        dispatch({ type: 'GENERIC_ERROR', error });
+    };
 
-      let boards = boardList
-        .filter(row => row['matching_boards']?.find(o => fqbn_list.includes(o['fqbn'])))
-        .map(board => ({
-          name: board['matching_boards'][0]["name"],
-          fqbn: board['matching_boards'][0]["fqbn"],
-          port: board['port']["address"],
-        }))
+    useEffect(() => {
+        textAreaRef.current.scrollTop = textAreaRef.current.scrollHeight;
+    }, [state.stdout]);
 
-      setBoards(boards)
-    }
+    useEffect(() => {
+        const getVersion = async () => {
```

Comment:
I am wondering if it will make sense to move all the functions interacting with the `arduino-cli` on a different file, something like `arduino_service.js` (I am not good with names though)

Reply:
It totally makes sense. Please have a look and let me know what do you think. I took the opportunity also to split the logic in App.jsx into multiple React components.